### ISSUE TITLE
[@types/jquery] .get() should return HTMLElement | undefined

### ISSUE THREAD

**Issue by ethanredmond2** [DefinitelyTyped/DefinitelyTyped#52052]
JQuery's [`.get()`](https://api.jquery.com/get/) method reports that it always returns Element, however in practice we can see that this is not always true:

```
$('#asdf').get(0).tagName
```

```
Uncaught TypeError: Cannot read property 'tagName' of undefined
```

To fix this, you could change the return type of `.get()` to `HTMLElement | undefined`. This may cause warnings when other projects update their version of @types/jquery, however I believe it'll be better for preventing possible errors.

TypeScript's `strictNullChecks` setting is specifically designed to warn about this kind of problem, for example, `document.getElementById()` returns `HTMLElement | undefined`. This encourages safer coding practices like [optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining), e.g.  `document.getElementById('asdf')?.focus()`

@leonard-thieu @borisyankov @choffmeister @Diullei @tasoili @jasons-novaleaf @seanski @Guuz @ksummerlin @basarat @nwolverson
@derekcicerone @AndrewGaspar @seikichi @benjaminjackman @s093294 @JoshStrobl @johnnyreilly @DickvdBrink @King2500 @terrymun @martin-badin

<sub>(I'm using TypeScript 3.9.9 and @types/jquery 3.5.5)</sub>

**Comment by martin-badin** [comment]
Please. Can you try?

$<HTMLElement>('#asdf').get(0).tagName

**Comment by ethanredmond2** [comment]
@martin-badin Hi Martin, I'm sorry, I don't understand how casting `.get(0)` to HTMLElement will solve the problem that TypeScript doesn't warn that the return value might be undefined, since `.get(0)` already returns type HTMLElement. Thanks

**Comment by martin-badin** [comment]
I'm sorry, I misunderstood. By the documentation is the return type correct https://api.jquery.com/get/. I know what you have in your mind but I think that your example/code is wroten wrong. Your code should be like this because `$('#asdf')` can return HTMLElement or undefined.
```
if($('#asdf').length) {
  $('#asdf').get(0).tagName
}
```

**Comment by ethanredmond2** [comment]
Yes, to prevent errors where there's no element with ID `asdf`, I could write:

```
var $asdf = $('#asdf');
if ($asdf.length) {
    $asdf.get(0).tagName
}
```
Or equally:
```
$('#asdf').get(0)?.tagName
```

I'm filing this issue because `.get(0)` can indeed return undefined, but @types/jquery doesn't acknowledge this, saying instead that `.get(0)` can _only_ return HTMLElement. This is incorrect because when there's no element with ID `asdf` and you call `$('#asdf').get(0)`, it does return undefined, so to make the types more accurate, the return type for `.get(0)` should be HTMLElement | undefined.

**Comment by miken32** [comment]
@martin-badin FYI, documentation you linked to states "If the value of `index` is out of bounds — less than the negative number of elements or equal to or greater than the number of elements — it returns `undefined`."

**Comment by princefishthrower** [comment]
@ethanredmond2 - I agree with you - it's exactly what the docs state. I've made a pull request for this change: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56118

**Comment by HolgerJeromin** [comment]
I think we can close this issue as solved.

**Comment by peterblazejewicz** [comment]
fixed in #56118
